File: crm-backend/crm-core/src/main/java/com/opticrm/core/lead/entity/Lead.java
package com.opticrm.core.lead.entity;

import com.opticrm.core.account.entity.Account;
import com.opticrm.core.contact.entity.Contact;
import com.opticrm.security.entity.Territory;
import com.opticrm.security.entity.User;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.UuidGenerator;
import org.springframework.data.annotation.CreatedBy;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.Instant;
import java.util.List;
import java.util.UUID;
import java.time.LocalDate;

@Entity
@Table(name = "leads")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class Lead {

    @Id
    @UuidGenerator
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    @Column(name = "salutation", length = 20)
    private String salutation;

    @Column(name = "first_name", length = 100)
    private String firstName;

    @Column(name = "last_name", nullable = false, length = 100)
    private String lastName;

    @Column(name = "email", length = 255)
    private String email;

    @Column(name = "phone", length = 20)
    private String phone;

    @Column(name = "mobile", length = 20)
    private String mobile;

    @Column(name = "company_name", length = 255)
    private String companyName;

    @Column(name = "job_title", length = 100)
    private String jobTitle;

    @Column(name = "website", length = 255)
    private String website;

    // Address
    @Column(name = "street", length = 255)
    private String street;

    @Column(name = "city", length = 100)
    private String city;

    @Column(name = "state", length = 100)
    private String state;

    @Column(name = "postal_code", length = 20)
    private String postalCode;

    @Column(name = "country", length = 100)
    private String country;

    // Lead Info
    @Column(name = "status", nullable = false, length = 50)
    @Builder.Default
    private String status = "new";

    @Column(name = "source", length = 50)
    private String source;

    @Column(name = "source_details", length = 255)
    private String sourceDetails;

    @Column(name = "rating", length = 20)
    private String rating;

    @Column(name = "lead_score")
    @Builder.Default
    private Integer leadScore = 0;

    @Column(name = "interested_products", columnDefinition = "TEXT[]")
    private List<String> interestedProducts;

    @Column(name = "budget_range", length = 50)
    private String budgetRange;

    @Column(name = "decision_timeframe", length = 50)
    private String decisionTimeframe;

    // Champs avancés
    @Column(name = "fax", length = 20)
    private String fax;

    @Column(name = "linkedin_url", length = 255)
    private String linkedinUrl;

    @Column(name = "industry", length = 100)
    private String industry;

    @Column(name = "num_employees", length = 50)
    private String numEmployees;

    @Column(name = "annual_revenue", length = 50)
    private String annualRevenue;

    @Column(name = "priority", length = 20)
    private String priority;

    @Column(name = "do_not_call")
    @Builder.Default
    private Boolean doNotCall = false;

    @Column(name = "do_not_email")
    @Builder.Default
    private Boolean doNotEmail = false;

    @Column(name = "next_follow_up_date")
    private LocalDate nextFollowUpDate;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    // Qualification avancée (V26)
    @Column(name = "qualification_notes", columnDefinition = "TEXT")
    private String qualificationNotes;

    @Column(name = "competitor", length = 200)
    private String competitor;

    @Column(name = "product_demo_requested")
    @Builder.Default
    private Boolean productDemoRequested = false;

    @Column(name = "meeting_scheduled_at")
    private Instant meetingScheduledAt;

    // Relations
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to_id")
    private User assignedTo;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "territory_id")
    private Territory territory;

    @Column(name = "campaign_id")
    private UUID campaignId;

    // Conversion
    @Column(name = "is_converted")
    @Builder.Default
    private Boolean isConverted = false;

    @Column(name = "converted_at")
    private Instant convertedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "converted_contact_id")
    private Contact convertedContact;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "converted_account_id")
    private Account convertedAccount;

    @Column(name = "converted_opportunity_id")
    private UUID convertedOpportunityId;

    // Optimistic locking
    @Version
    @Column(name = "version")
    private Long version;

    // Audit
    @CreatedBy
    @Column(name = "created_by_id", updatable = false)
    private UUID createdById;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private Instant updatedAt;

    @Column(name = "last_activity_at")
    private Instant lastActivityAt;

    // Helpers
    public String getFullName() {
        StringBuilder sb = new StringBuilder();
        if (salutation != null && !salutation.isEmpty()) {
            sb.append(salutation).append(" ");
        }
        if (firstName != null) {
            sb.append(firstName).append(" ");
        }
        sb.append(lastName);
        return sb.toString().trim();
    }

    public String getDisplayName() {
        if (firstName != null && !firstName.isEmpty()) {
            return firstName + " " + lastName;
        }
        return lastName;
    }

    public String getFullAddress() {
        StringBuilder sb = new StringBuilder();
        if (street != null) sb.append(street);
        if (postalCode != null || city != null) {
            if (sb.length() > 0) sb.append(", ");
            if (postalCode != null) sb.append(postalCode).append(" ");
            if (city != null) sb.append(city);
        }
        if (country != null) {
            if (sb.length() > 0) sb.append(", ");
            sb.append(country);
        }
        return sb.toString();
    }
}

File: crm-web/src/types/lead.ts
export interface Lead {
  id: string;
  salutation?: string;
  firstName?: string;
  lastName: string;
  fullName: string;
  email?: string;
  phone?: string;
  mobile?: string;
  companyName?: string;
  jobTitle?: string;
  website?: string;

  // Address
  street?: string;
  city?: string;
  state?: string;
  postalCode?: string;
  country?: string;
  fullAddress?: string;

  // Lead Info
  status: string;
  source?: string;
  sourceDetails?: string;
  rating?: string;
  leadScore?: number;
  interestedProducts?: string[];
  budgetRange?: string;
  decisionTimeframe?: string;

  // Champs avancés
  fax?: string;
  linkedinUrl?: string;
  industry?: string;
  numEmployees?: string;
  annualRevenue?: string;
  priority?: string;
  doNotCall?: boolean;
  doNotEmail?: boolean;
  nextFollowUpDate?: string;
  description?: string;

  // Qualification avancée (V26)
  qualificationNotes?: string;
  competitor?: string;
  productDemoRequested?: boolean;
  meetingScheduledAt?: string;

  // Relations
  assignedTo?: {
    id: string;
    fullName: string;
    email: string;
  };
  territory?: {
    id: string;
    name: string;
    region?: string;
  };
  campaignId?: string;

  // Conversion
  isConverted?: boolean;
  convertedAt?: string;
  convertedContactId?: string;
  convertedAccountId?: string;
  convertedOpportunityId?: string;

  // Audit
  createdAt: string;
  updatedAt?: string;
  lastActivityAt?: string;
}

export interface LeadListItem {
  id: string;
  firstName?: string;
  lastName: string;
  fullName: string;
  email?: string;
  phone?: string;
  companyName?: string;
  jobTitle?: string;
  city?: string;
  country?: string;
  status: string;
  source?: string;
  rating?: string;
  leadScore?: number;
  assignedToId?: string;
  assignedToName?: string;
  createdAt: string;
  lastActivityAt?: string;
}

export interface CreateLeadRequest {
  salutation?: string;
  firstName?: string;
  lastName: string;
  email?: string;
  phone?: string;
  mobile?: string;
  companyName?: string;
  jobTitle?: string;
  website?: string;
  street?: string;
  city?: string;
  state?: string;
  postalCode?: string;
  country?: string;
  status?: string;
  source?: string;
  sourceDetails?: string;
  rating?: string;
  leadScore?: number;
  interestedProducts?: string[];
  budgetRange?: string;
  decisionTimeframe?: string;
  fax?: string;
  linkedinUrl?: string;
  industry?: string;
  numEmployees?: string;
  annualRevenue?: string;
  priority?: string;
  doNotCall?: boolean;
  doNotEmail?: boolean;
  nextFollowUpDate?: string;
  description?: string;
  qualificationNotes?: string;
  competitor?: string;
  productDemoRequested?: boolean;
  meetingScheduledAt?: string;
  assignedToId?: string;
  territoryId?: string;
  campaignId?: string;
}

export interface UpdateLeadRequest extends Partial<CreateLeadRequest> {}

export interface LeadScoreBreakdown {
  profile: number;
  qualification: number;
  engagement: number;
  hasEmail: boolean;
  hasPhone: boolean;
  hasCompany: boolean;
  hasJobTitle: boolean;
  hasCity: boolean;
  hasWebOrLinkedin: boolean;
  hasBudgetRange: boolean;
  hasDecisionTimeframe: boolean;
  hasInterestedProducts: boolean;
  hasIndustry: boolean;
  hasNumEmployees: boolean;
  status: string;
  rating: string;
  hasMeetingScheduled: boolean;
  hasDemoRequested: boolean;
}

export interface LeadScoreDto {
  score: number;
  label: string;
  color: string;
  breakdown: LeadScoreBreakdown;
}

export interface ConvertLeadRequest {
  createAccount?: boolean;
  createContact?: boolean;
  createOpportunity?: boolean;
  accountName?: string;
  accountType?: string;
  industryId?: string;
  opportunityName?: string;
  opportunityAmount?: number;
  opportunityStageId?: string;
  closeDate?: string;
  existingAccountId?: string;
}

export const SALUTATIONS = [
  { value: 'M.', label: 'M.' },
  { value: 'Mme', label: 'Mme' },
  { value: 'Dr.', label: 'Dr.' },
  { value: 'Pr.', label: 'Pr.' },
  { value: 'Me', label: 'Me' },
];

export const LEAD_STATUSES = [
  { value: 'new', label: 'Nouveau', color: 'blue' },
  { value: 'contacted', label: 'Contacté', color: 'cyan' },
  { value: 'qualified', label: 'Qualifié', color: 'green' },
  { value: 'unqualified', label: 'Non qualifié', color: 'red' },
  { value: 'nurturing', label: 'En nurturing', color: 'purple' },
  { value: 'converted', label: 'Converti', color: 'gold' },
];

export const LEAD_SOURCES = [
  { value: 'website', label: 'Site web' },
  { value: 'referral', label: 'Recommandation' },
  { value: 'cold_call', label: 'Appel à froid' },
  { value: 'trade_show', label: 'Salon' },
  { value: 'social_media', label: 'Réseaux sociaux' },
  { value: 'advertising', label: 'Publicité' },
  { value: 'email_campaign', label: 'Campagne email' },
  { value: 'partner', label: 'Partenaire' },
  { value: 'other', label: 'Autre' },
];

export const LEAD_RATINGS = [
  { value: 'hot', label: 'Chaud', color: 'red' },
  { value: 'warm', label: 'Tiède', color: 'orange' },
  { value: 'cold', label: 'Froid', color: 'blue' },
];

export const BUDGET_RANGES = [
  { value: 'under_10k', label: '< 10 000 €' },
  { value: '10k_50k', label: '10 000 € - 50 000 €' },
  { value: '50k_100k', label: '50 000 € - 100 000 €' },
  { value: '100k_500k', label: '100 000 € - 500 000 €' },
  { value: 'over_500k', label: '> 500 000 €' },
];

export const DECISION_TIMEFRAMES = [
  { value: 'immediate', label: 'Immédiat' },
  { value: '1_month', label: 'Dans le mois' },
  { value: '1_3_months', label: '1 à 3 mois' },
  { value: '3_6_months', label: '3 à 6 mois' },
  { value: '6_12_months', label: '6 à 12 mois' },
  { value: 'over_1_year', label: 'Plus d\'un an' },
];

export const LEAD_PRIORITIES = [
  { value: 'high', label: 'Haute', color: 'red' },
  { value: 'medium', label: 'Moyenne', color: 'orange' },
  { value: 'low', label: 'Basse', color: 'blue' },
];

export const INDUSTRIES = [
  { value: 'agriculture', label: 'Agriculture' },
  { value: 'btp', label: 'BTP / Construction' },
  { value: 'commerce', label: 'Commerce & Distribution' },
  { value: 'education', label: 'Éducation & Formation' },
  { value: 'energie', label: 'Énergie & Environnement' },
  { value: 'finance', label: 'Finance & Banque' },
  { value: 'immobilier', label: 'Immobilier' },
  { value: 'industrie', label: 'Industrie & Manufacture' },
  { value: 'informatique', label: 'Informatique & Tech' },
  { value: 'sante', label: 'Santé & Pharmacie' },
  { value: 'services', label: 'Services aux entreprises' },
  { value: 'telecom', label: 'Télécommunications' },
  { value: 'tourisme', label: 'Tourisme & Hôtellerie' },
  { value: 'transport', label: 'Transport & Logistique' },
  { value: 'autre', label: 'Autre' },
];

export const NUM_EMPLOYEES_RANGES = [
  { value: '1_10', label: '1 – 10' },
  { value: '11_50', label: '11 – 50' },
  { value: '51_200', label: '51 – 200' },
  { value: '201_500', label: '201 – 500' },
  { value: '501_1000', label: '501 – 1 000' },
  { value: 'over_1000', label: '+ 1 000' },
];

export const ANNUAL_REVENUES = [
  { value: 'under_1m', label: '< 1 M MAD' },
  { value: '1m_5m', label: '1 – 5 M MAD' },
  { value: '5m_20m', label: '5 – 20 M MAD' },
  { value: '20m_100m', label: '20 – 100 M MAD' },
  { value: 'over_100m', label: '> 100 M MAD' },
];
